File: apps/backend/src/scopes/scope-parser.ts
import fs from 'fs/promises';
import path from 'path';
import matter from 'gray-matter';

export interface ScopeMetadata {
  scope_id: string;
  title: string;
  phase: 'analyze' | 'design' | 'build' | 'test' | 'deploy' | 'monitor' | 'iterate';
  status: 'open' | 'active' | 'done' | 'blocked' | 'conflict';
  agent_id?: string;
  created_at: string;
  updated_at: string;
  priority: 'low' | 'normal' | 'high';
}

export interface Scope {
  metadata: ScopeMetadata;
  description: string;
  inputs: string;
  outputs: string;
  memory: string;
  result: string;
}

export class ScopeParser {
  static parse(content: string): Scope {
    const { data, content: body } = matter(content);

    const sections = this.extractSections(body);

    return {
      metadata: data as ScopeMetadata,
      description: sections.description || '',
      inputs: sections.inputs || '',
      outputs: sections.outputs || '',
      memory: sections.memory || '',
      result: sections.result || ''
    };
  }

  static serialize(scope: Scope): string {
    const frontmatter = matter.stringify('', scope.metadata);
    const body = `
## Description

${scope.description}

## Inputs

${scope.inputs}

## Outputs

${scope.outputs}

## Memory / Working Notes

${scope.memory}

## Result

${scope.result}
`;
    return frontmatter.trim() + '\n' + body.trim() + '\n';
  }

  private static extractSections(content: string): Record<string, string> {
    const sections: Record<string, string> = {};
    const sectionRegex = /^## (.+)$/gm;

    const matches = [...content.matchAll(sectionRegex)];

    for (let i = 0; i < matches.length; i++) {
      const match = matches[i];
      const sectionName = match[1].toLowerCase().replace(/\s+/g, '_').replace(/\//g, '_');
      const startIndex = match.index! + match[0].length;
      const endIndex = i < matches.length - 1 ? matches[i + 1].index! : content.length;
      sections[sectionName] = content.substring(startIndex, endIndex).trim();
    }

    return sections;
  }
}

File: apps/backend/Services/ScopeParser.cs
using System.Text;
using System.Text.RegularExpressions;
using YamlDotNet.Serialization;
using YamlDotNet.Serialization.NamingConventions;
using ControlPlane.Backend.Models;

namespace ControlPlane.Backend.Services;

public class ScopeParser
{
    private static readonly ISerializer _yamlSerializer = new SerializerBuilder()
        .WithNamingConvention(UnderscoredNamingConvention.Instance)
        .Build();

    private static readonly IDeserializer _yamlDeserializer = new DeserializerBuilder()
        .WithNamingConvention(UnderscoredNamingConvention.Instance)
        .IgnoreUnmatchedProperties()
        .Build();

    public static Scope Parse(string content)
    {
        var parts = content.Split(new[] { "---" }, StringSplitOptions.None);

        if (parts.Length < 3)
        {
            throw new InvalidOperationException("Invalid markdown format");
        }

        var yamlContent = parts[1].Trim();
        var bodyContent = parts[2].Trim();

        var metadata = _yamlDeserializer.Deserialize<ScopeMetadata>(yamlContent);
        var sections = ExtractSections(bodyContent);

        return new Scope
        {
            metadata = metadata,
            description = sections.GetValueOrDefault("description", ""),
            inputs = sections.GetValueOrDefault("inputs", ""),
            outputs = sections.GetValueOrDefault("outputs", ""),
            memory = sections.GetValueOrDefault("memory_working_notes", ""),
            result = sections.GetValueOrDefault("result", "")
        };
    }

    public static string Serialize(Scope scope)
    {
        var yaml = _yamlSerializer.Serialize(scope.metadata);

        var body = $@"
## Description

{scope.description}

## Inputs

{scope.inputs}

## Outputs

{scope.outputs}

## Memory / Working Notes

{scope.memory}

## Result

{scope.result}
";

        return $"---\n{yaml.Trim()}\n---\n{body.Trim()}\n";
    }

    private static Dictionary<string, string> ExtractSections(string content)
    {
        var sections = new Dictionary<string, string>();
        var regex = new Regex(@"^## (.+)$", RegexOptions.Multiline);
        var matches = regex.Matches(content);

        for (int i = 0; i < matches.Count; i++)
        {
            var match = matches[i];
            var sectionName = match.Groups[1].Value
                .ToLower()
                .Replace(" ", "_")
                .Replace("/", "_");

            var startIndex = match.Index + match.Length;
            var endIndex = i < matches.Count - 1 ? matches[i + 1].Index : content.Length;

            sections[sectionName] = content.Substring(startIndex, endIndex - startIndex).Trim();
        }

        return sections;
    }
}
